pkg/workflow/controllers/chaos_node_reconciler.go

```diff
@@ -0,0 +1,205 @@
+	if availableChaos(string(node.Spec.Type)) {
+		err = it.injectChaos(ctx, node)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (it *ChaosNodeReconciler) injectChaos(ctx context.Context, node v1alpha1.WorkflowNode) error {
+	var chaosObject runtime.Object
+	var meta metav1.Object
+
+	if node.Spec.Type == v1alpha1.TypeNetworkChaos {
+		networkChaos := v1alpha1.NetworkChaos{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", node.Name),
+				Namespace:    node.Namespace,
+			},
+			Spec: *node.Spec.NetworkChaos,
+		}
+		meta = networkChaos.GetObjectMeta()
+		chaosObject = &networkChaos
+	} else if node.Spec.Type == v1alpha1.TypePodChaos {
+		podChaos := v1alpha1.PodChaos{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", node.Name),
+				Namespace:    node.Namespace,
+			},
+			Spec: *node.Spec.PodChaos,
+		}
+		meta = podChaos.GetObjectMeta()
+		chaosObject = &podChaos
+	} else {
+		it.logger.Info("unsupported chaos nodes", "key", fmt.Sprintf("%s/%s", node.Namespace, node.Name), "type", node.Spec.Type)
+		return nil
+	}
+
+	meta.SetOwnerReferences(append(meta.GetOwnerReferences(), metav1.OwnerReference{
+		APIVersion:         node.APIVersion,
+		Kind:               node.Kind,
+		Name:               node.Name,
+		UID:                node.UID,
+		Controller:         &isController,
+		BlockOwnerDeletion: &blockOwnerDeletion,
+	}))
+
+	err := it.kubeClient.Create(ctx, chaosObject)
+	if err != nil {
+		it.eventRecorder.Event(&node, corev1.EventTypeWarning, v1alpha1.ChaosCRCreateFailed, "Failed to create chaos CR")
+		it.logger.Error(err, "failed to create chaos")
+		return nil
+	}
+
+	it.eventRecorder.Event(&node, corev1.EventTypeNormal, v1alpha1.ChaosCRCreated, fmt.Sprintf("Chaos CR %s/%s created", meta.GetNamespace(), meta.GetName()))
+
+	group := chaosObject.GetObjectKind().GroupVersionKind().Group
+	chaosRef := corev1.TypedLocalObjectReference{
+		APIGroup: &group,
+		Kind:     chaosObject.GetObjectKind().GroupVersionKind().Kind,
+		Name:     meta.GetName(),
+	}
+
+	updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		node := v1alpha1.WorkflowNode{}
+		err := it.kubeClient.Get(ctx, types.NamespacedName{
+			Namespace: node.Namespace,
+			Name:      node.Name,
+		}, &node)
+		if err != nil {
+			return client.IgnoreNotFound(err)
+		}
+		node.Status.ChaosResource = &chaosRef
+
+		// TODO: this condition should be set by observation
+		SetCondition(&node.Status, v1alpha1.WorkflowNodeCondition{
+			Type:   v1alpha1.ConditionChaosInjected,
+			Status: corev1.ConditionTrue,
+			Reason: v1alpha1.ChaosCRCreated,
+		})
+
+		return it.kubeClient.Update(ctx, &node)
+
+	})
+	return updateError
+}
+
+func (it *ChaosNodeReconciler) recoverChaos(ctx context.Context, node v1alpha1.WorkflowNode) error {
+	if node.Status.ChaosResource == nil {
+		return nil
+	}
+
+	var chaosObject runtime.Object
+
+	var err error
+	if node.Spec.Type == v1alpha1.TypeNetworkChaos {
```

Comment:
Seems there're tons of code to be generated. But I fear that code generating would not be the best practice to do such things.

Reply:
> Seems there're tons of code to be generated.

Yeah, I suppose that we will use code generation in the future. I will finish that in the next PRs.

>But I fear that code generating would not be the best practice to do such things.

Do you have any suggestions?

Reply:
I'm not sure if we can actually do this in go, so  it's just a thought.
Is there any chance we can do something like
```go
type = some_match_function(node.Spec.Type); // yields some reflect.Type
chaos = reflect.New(type); // and fill in the specs
```
there will still be some code generation to do in `some_match_function` but that will be less duplication.

Reply:
Yeah, we could do that, but it still needs code generation in `some_match_function()`. I will compare these different codes and choose the sipmler one.

Reply:
In fact I was wondering if there should be one and only one way to apply transition between a `string` and some corresponding `reflect.Type`. In fact there're already different cases in the code now. If I haven't got it wrong, one is here 
https://github.com/chaos-mesh/chaos-mesh/blob/c9d9f0706ae1c1c1ee0cc06d3df6b5370ff5505b/pkg/apiserver/experiment/experiment.go#L147-L155
and 
https://github.com/chaos-mesh/chaos-mesh/blob/c9d9f0706ae1c1c1ee0cc06d3df6b5370ff5505b/pkg/apiserver/experiment/experiment.go#L812-L827
also
https://github.com/chaos-mesh/chaos-mesh/blob/c9d9f0706ae1c1c1ee0cc06d3df6b5370ff5505b/cmd/chaos-builder/validator.go#L21-L32

Reply:
Yeah, I think we could reuse the `Kind<type>Chaos` and `chaosKindMap`. 

But I think I will write another (duplicated) code generation for the type transition things, and make refactor to "one and only one way for `string`/`reflect.Type` transition" after that.

Reply:
FYI @WangXiangUSTC

Reply:
In fact, I always prefer code generation than `reflect` (if code generation is possible). But code generator is really hard to write. I wonder whether there is something like [`quote!`](https://github.com/dtolnay/quote) in the go world? (I think go ecosystem depends heavily on the code generation)